code/io/tree_hdf5.c

```diff
@@ -0,0 +1,513 @@
+  }
+
+  status = fill_metadata_names(&metadata_names, TreeType);
+  if (status != EXIT_SUCCESS)
+  {
+    ABORT(0);
+  }
+ 
+  status = read_attribute_int(hdf5_file, "/Header", metadata_names.name_NTrees, &Ntrees);
+  if (status != EXIT_SUCCESS)
+  {
+    fprintf(stderr, "Error while processing file %s\n", buf);
+    fprintf(stderr, "Error code is %d\n", status);
+    ABORT(0);
+  }
+
+  status = read_attribute_int(hdf5_file, "/Header", metadata_names.name_totNHalos, &totNHalos);
+  if (status != EXIT_SUCCESS)  
+  {
+    fprintf(stderr, "Error while processing file %s\n", buf);
+    fprintf(stderr, "Error code is %d\n", status);
+    ABORT(0);
+  }
+
+  printf("There are %d trees and %d total halos\n", Ntrees, totNHalos);
+  
+  TreeNHalos = mymalloc(sizeof(int) * Ntrees); 
+
+  status = read_attribute_int(hdf5_file, "/Header", metadata_names.name_TreeNHalos, TreeNHalos);
+  if (status != EXIT_SUCCESS)
+  {
+    fprintf(stderr, "Error while processing file %s\n", buf);
+    fprintf(stderr, "Error code is %d\n", status);
+    ABORT(0);
+  }
+
+  TreeFirstHalo = mymalloc(sizeof(int) * Ntrees);
+
+  for (i = 0; i < 20; ++i)
+    printf("Tree %d: NHalos %d\n", i, TreeNHalos[i]);
+
+  if(Ntrees)
+    TreeFirstHalo[0] = 0;
+  for(i = 1; i < Ntrees; i++)
+    TreeFirstHalo[i] = TreeFirstHalo[i - 1] + TreeNHalos[i - 1];
+
+ // H5Fclose(hdf5_file);
+
+}
+
+#define READ_TREE_FIELD(file, name, type, h5type)\
+{\
+  hid_t dataset_id = H5Dopen(file, #name, H5P_DEFAULT);\
+  H5Dread(dataset_id, h5type, H5S_ALL, H5S_ALL, H5P_DEFAULT, buffer);\
+  for (int ii = 0; ii < NHalos_ThisTree; ++ii)\
+  {\
+    Halo[ii].name = ((type*)buffer)[ii];\
+  }\
+}
+
+void load_tree_hdf5(int32_t filenr, int32_t treenr)
+{
+
+  char dataset_name[MAX_STRING_LEN];
+  int32_t NHalos_ThisTree, status, halo_idx, dim;
+
+  int32_t *buffer_int;
+  float *buffer_float;
+  long long *buffer_longlong;
+
+  if (hdf5_file <= 0)
+  {
+    fprintf(stderr, "The HDF5 file should still be opened when reading the halos in the tree.\n");
+    fprintf(stderr, "For tree %d we encountered error %d\n", treenr, hdf5_file);
+    ABORT(0);
+  }
+
+  NHalos_ThisTree = TreeNHalos[treenr];
+
+  Halo = mymalloc(sizeof(struct halo_data) * NHalos_ThisTree); 
+
+  // We now need to read in all the halo fields for this tree.
+  // To do so, we read the field into a buffer and then properly slot the field into the Halo struct.
+
+
+  // TODO: This is currently a very (VERY) messy way of doing things.
+  // Ideally we would just like to specify the field name of the HDF5 file and the corresponding field in the Halo struct.
+  // However it's a bit tricky because we would need to account for the case of Position where the HDF5 dataset is Nx3.
+  // Furthermore, also need to think about how to do the slotting into the Halo struct.  I don't think this is possibly to do all at once.
+
+  /* Merger Tree Pointers */ 
+
+  // Descendant //
+  snprintf(dataset_name, MAX_STRING_LEN - 1, "tree_%03d/Descendant", treenr);
```

Comment:
Excellent `snprintf` :)

Reply:
I think you could wrap `snprintf` into a macro, because typing in `MAX_STRING_LEN - 1` might get annoying